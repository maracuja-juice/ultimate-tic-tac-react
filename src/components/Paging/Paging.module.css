.pagingButton {
  min-width: 30px;
  padding: 0.3em;
}

.previousButton {
  margin-right: 0.8em;
}

.nextButton {
  margin-left: 0.8em;
}
