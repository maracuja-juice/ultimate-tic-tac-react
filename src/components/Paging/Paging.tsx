import * as React from 'react';
import { Button } from '@rmwc/button';
import styles from './Paging.module.css';
import { useState } from 'react';

interface PagingProps {
  pages: number;
  pageToStartWith: number;
  onPageChange: (/* lastPage: number, */ newPage: number) => void;
  pagesToDisplayFrom?: number;
  pagesToDisplayTo?: number;
}

export function Paging(props: PagingProps) {
  const { pages, onPageChange } = props;

  const [currentPage, setCurrentPage] = useState(props.pageToStartWith);

  const getPages = () => {
    const pageButtons: JSX.Element[] = [];
    for (let i = 1; i < pages + 1; i++) {
      pageButtons.push(
        <Button
          key={i}
          label={i}
          onClick={() => {
            if (i !== currentPage) {
              setCurrentPage(i);
              onPageChange(i);
            }
          }}
          outlined={i === currentPage}
          dense={true}
          ripple={false}
          className={styles.pagingButton}
        />,
      );
    }
    return pageButtons;
  };

  return (
    <div>
      <Button
        label={'Previous'}
        onClick={() => {
          const newPage = currentPage - 1;
          setCurrentPage(newPage);
          onPageChange(newPage);
        }}
        disabled={currentPage === 1}
        icon={{ icon: 'arrow-left', 'aria-hidden': true }}
        ripple={false}
        dense={true}
        className={styles.previousButton}
      />
      {getPages()}
      <Button
        label={'Next'}
        onClick={() => {
          const newPage = currentPage + 1;
          setCurrentPage(newPage);
          onPageChange(newPage);
        }}
        disabled={currentPage === pages}
        trailingIcon={{ icon: 'arrow-right', 'aria-hidden': true }}
        ripple={false}
        dense={true}
        className={styles.nextButton}
      />
    </div>
  );
}
