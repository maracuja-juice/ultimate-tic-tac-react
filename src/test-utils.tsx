// See: https://testing-library.com/docs/react-testing-library/setup#custom-render
import React, { FC, ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { IconProvider } from './components/Icons';
import '@testing-library/jest-dom'; // https://github.com/testing-library/jest-dom/issues/282#issuecomment-660034778
import { MemoryRouter } from 'react-router-dom';

const AllTheProviders: FC = ({ children }) => {
  return (
    <IconProvider>
      <MemoryRouter>{children}</MemoryRouter>
    </IconProvider>
  );
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';

export { customRender as render };
