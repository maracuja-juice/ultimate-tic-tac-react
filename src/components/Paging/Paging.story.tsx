import * as React from 'react';
import { Paging, PagingProps } from './Paging';
import { action } from '@storybook/addon-actions';
import { Story } from '@storybook/react/types-6-0';
import { IconProvider } from '../IconProvider';

export default {
  title: 'Paging',
};

export const Standard: Story<PagingProps> = (args) => (
  <IconProvider>
    <Paging {...args} onPageChange={action('onPageChange')} />
  </IconProvider>
);
Standard.args = {
  pages: 10,
  pageToStartWith: 10
};