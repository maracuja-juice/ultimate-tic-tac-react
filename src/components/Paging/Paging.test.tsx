import * as React from 'react';
// TODO make this import cleaner
// -> like this probably: https://create-react-app.dev/docs/importing-a-component/#absolute-imports
import { render, screen } from '../../test-utils';
import userEvent from '@testing-library/user-event';
import { composeStories } from '@storybook/testing-react';
import * as stories from './Paging.story';
const { Standard, Empty, Minimum, Maximum } = composeStories(stories);

test('renders supplied amount of page buttons', () => {
  render(<Standard />);
  const buttons = screen.getAllByRole('button', { name: /\d/ });
  expect(buttons).toHaveLength(10);
});

test('renders no buttons when pages are 0', () => {
  render(<Empty />);
  const buttons = screen.queryAllByRole('button');
  expect(buttons).toHaveLength(0);
});

test('allows using next button', () => {
  const onPageChange = jest.fn();
  render(<Standard onPageChange={onPageChange} />);

  const next = screen.getByRole('button', { name: /next/i });

  const pageSix = screen.getByRole('button', { name: /6/ });
  expect(pageSix.className).not.toMatch(/unelevated/);

  userEvent.click(next);

  expect(pageSix.className).toMatch(/unelevated/);
  expect(onPageChange).toHaveBeenCalledWith(6);
});

test('allows using previous button', () => {
  const onPageChange = jest.fn();
  render(<Standard onPageChange={onPageChange} />);

  const previous = screen.getByRole('button', { name: /previous/i });

  const pageFour = screen.getByRole('button', { name: /4/ });
  expect(pageFour.className).not.toMatch(/unelevated/);

  userEvent.click(previous);

  expect(pageFour.className).toMatch(/unelevated/);
  expect(onPageChange).toHaveBeenCalledWith(4);
});

test('allows using specific page button', () => {
  const onPageChange = jest.fn();
  render(<Standard onPageChange={onPageChange} />);

  const page = screen.getByRole('button', { name: /7/ });
  expect(page.className).not.toMatch(/unelevated/);
  userEvent.click(page);

  expect(onPageChange).toHaveBeenCalledWith(7);
  expect(page.className).toMatch(/unelevated/);
});

test('doesnt allow previous button on Minimum', () => {
  const onPageChange = jest.fn();
  render(<Minimum onPageChange={onPageChange} />);

  const previous = screen.getByRole('button', { name: /previous/i });
  userEvent.click(previous);

  expect(onPageChange).not.toHaveBeenCalled();
});

test('doesnt allow next button on Maximum', () => {
  const onPageChange = jest.fn();
  render(<Maximum onPageChange={onPageChange} />);

  const next = screen.getByRole('button', { name: /next/i });
  userEvent.click(next);

  expect(onPageChange).not.toHaveBeenCalled();
});
