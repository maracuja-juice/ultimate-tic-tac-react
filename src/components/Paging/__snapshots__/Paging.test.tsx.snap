// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paging should match snapshot 1`] = `
<div>
  <ArrowButtons
    maxValue={10}
    minValue={1}
    onInteraction={[Function]}
    value={1}
  >
    <withRipple(Function)
      className="pagingButton"
      dense={true}
      key="1"
      label={1}
      onClick={[Function]}
      unelevated={true}
    />
    <withRipple(Function)
      className="pagingButton"
      dense={true}
      key="2"
      label={2}
      onClick={[Function]}
      unelevated={false}
    />
    <withRipple(Function)
      className="pagingButton"
      dense={true}
      key="3"
      label={3}
      onClick={[Function]}
      unelevated={false}
    />
    <withRipple(Function)
      className="pagingButton"
      dense={true}
      key="4"
      label={4}
      onClick={[Function]}
      unelevated={false}
    />
    <withRipple(Function)
      className="pagingButton"
      dense={true}
      key="5"
      label={5}
      onClick={[Function]}
      unelevated={false}
    />
    <withRipple(Function)
      className="pagingButton"
      dense={true}
      key="6"
      label={6}
      onClick={[Function]}
      unelevated={false}
    />
    <withRipple(Function)
      className="pagingButton"
      dense={true}
      key="7"
      label={7}
      onClick={[Function]}
      unelevated={false}
    />
    <withRipple(Function)
      className="pagingButton"
      dense={true}
      key="8"
      label={8}
      onClick={[Function]}
      unelevated={false}
    />
    <withRipple(Function)
      className="pagingButton"
      dense={true}
      key="9"
      label={9}
      onClick={[Function]}
      unelevated={false}
    />
    <withRipple(Function)
      className="pagingButton"
      dense={true}
      key="10"
      label={10}
      onClick={[Function]}
      unelevated={false}
    />
  </ArrowButtons>
</div>
`;
